src/EndpointV2/Ruleset/Ruleset.php

```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace Aws\EndpointV2\Ruleset;
+
+use Aws\EndpointV2\Rule\RuleCreator;
+
+/**
+ * A collection of rules, parameter specs and a class of helper functions
+ * used to resolve either an endpoint or an error.
+ */
+Class Ruleset
+{
+    /** @var string */
+    private $version;
+
+    /** @var array */
+    private $parameters;
+
+    /** @var array */
+    private $rules;
+
+    /** @var RulesetStandardLibrary */
+    public $standardLibrary;
+
+    public function __construct(array $ruleset, array $partitions)
+    {
+        $this->version = $ruleset['version'];
+        $this->parameters = $this->createParameters($ruleset['parameters']);
+        $this->rules = $this->createRules($ruleset['rules']);
+        $this->standardLibrary = new RulesetStandardLibrary($partitions);
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getVersion()
+    {
+        return $this->version;
+    }
+
+    /**
+     * @return array
+     */
+    public function getParameters()
+    {
+        return $this->parameters;
+    }
+
+    /**
+     * @return array
+     */
+    public function getRules()
+    {
+        return $this->rules;
+    }
+
+    /**
+     * Evaluate the ruleset against the input parameters.
+     * Return the first rule the parameters match against.
+     *
+     * @return mixed
+     */
+    public function evaluate(array $inputParameters)
+    {
+        $this->validateInputParameters($inputParameters);
+
+        foreach($this->rules as $rule) {
+            // newly set parameters via "assign" cannot be shared between
+            // adjacent rules
+            $inputParametersCopy = $inputParameters;
+            $evaluation = $rule->evaluate($inputParametersCopy, $this->standardLibrary);
+            if (!is_null($evaluation)) {
+                return $evaluation;
+            }
+        }
+        return null;
```

Comment:
Consider changing all of these to false so we can do a strict evaluation check to know for sure if an endpoint rule has been run yet; otherwise it could wind up being null before it's invoked which seems more error prone